src/phoenix/datasets/evaluators/utils.py

```diff
@@ -0,0 +1,164 @@
+import functools
+import inspect
+from typing import Any, Callable, Optional
+
+from phoenix.datasets.types import (
+    AnnotatorKind,
+    CanAsyncEvaluate,
+    CanEvaluate,
+    EvaluationResult,
+    EvaluatorCallable,
+    Example,
+    ExperimentEvaluator,
+    ExperimentRun,
+    JSONSerializable,
+)
+
+
+def _unwrap_json(obj: JSONSerializable) -> JSONSerializable:
+    if isinstance(obj, dict):
+        if len(obj) == 1:
+            key = next(iter(obj.keys()))
+            output = obj[key]
+            assert isinstance(
+                output, (dict, list, str, int, float, bool, type(None))
+            ), "Output must be JSON serializable"
+            return output
+    return obj
+
+
+def _validate_signature(sig: inspect.Signature) -> None:
+    # Check that the wrapped function has a valid signature for use as an evaluator
+    # If it does not, raise an error to exit early before running evaluations
+    params = sig.parameters
+    valid_named_params = {"experiment_input", "output", "reference", "metadata"}
+    if len(params) == 0:
+        raise ValueError("Evaluation function must have at least one parameter.")
+    if len(params) > 1:
+        not_found = set(params) - valid_named_params
+        if not_found:
+            raise ValueError(
+                (
+                    f"Invalid parameter names in evaluation function: {', '.join(not_found)}. "
+                    "Parameters names for multi-argument functions must be "
+                    f"any of: {', '.join(valid_named_params)}."
+                )
+            )
+
+
+def _bind_signature(
+    sig: inspect.Signature, example: Example, experiment_run: ExperimentRun
+) -> inspect.BoundArguments:
+    if experiment_run.output:
+        output = experiment_run.output.result
+    else:
+        output = None
+    parameter_mapping = {
+        "experiment_input": example.input,
+        "output": output,
```

Comment:
This is super slick - with one caviat that the nesting feels slightly off

```python

def no_error(output):
    return 1.0 if output["result"].get("error") is None else 0.0

def has_results(output):
    results = output["result"].get("results")
    print(results)
    has_results = results is not None and len(results) > 0
    return 1.0 if has_results else 0.0
```

really comes down to how we plan on exposing "nested" results.

Reply:
good point, I've un-nested if the result has the `result` key

Reply:
Well, I think we actually have to fix this unilaterally. It's kinda a problem that the output doesn't match the shape of expected output data model

Reply:
The latest version should have this fixed